File: app/src/test/java/com/vrem/wifianalyzer/wifi/band/WiFiChannelsGHZ2Test.java
/*
 * WiFiAnalyzer
 * Copyright (C) 2019  VREM Software Development <VREMSoftwareDevelopment@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>
 */

package com.vrem.wifianalyzer.wifi.band;

import org.junit.Before;
import org.junit.Test;

import java.util.List;
import java.util.Locale;

import androidx.core.util.Pair;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

public class WiFiChannelsGHZ2Test {

    private WiFiChannelsGHZ2 fixture;

    @Before
    public void setUp() {
        fixture = new WiFiChannelsGHZ2();
    }

    @Test
    public void testIsInRange() {
        assertTrue(fixture.isInRange(2400));
        assertTrue(fixture.isInRange(2499));
    }

    @Test
    public void testIsNotInRange() {
        assertFalse(fixture.isInRange(2399));
        assertFalse(fixture.isInRange(2500));
    }

    @Test
    public void testGetWiFiChannelByFrequency() {
        assertEquals(1, fixture.getWiFiChannelByFrequency(2410).getChannel());
        assertEquals(1, fixture.getWiFiChannelByFrequency(2412).getChannel());
        assertEquals(1, fixture.getWiFiChannelByFrequency(2414).getChannel());

        assertEquals(6, fixture.getWiFiChannelByFrequency(2437).getChannel());
        assertEquals(7, fixture.getWiFiChannelByFrequency(2442).getChannel());

        assertEquals(13, fixture.getWiFiChannelByFrequency(2470).getChannel());
        assertEquals(13, fixture.getWiFiChannelByFrequency(2472).getChannel());
        assertEquals(13, fixture.getWiFiChannelByFrequency(2474).getChannel());

        assertEquals(14, fixture.getWiFiChannelByFrequency(2482).getChannel());
        assertEquals(14, fixture.getWiFiChannelByFrequency(2484).getChannel());
        assertEquals(14, fixture.getWiFiChannelByFrequency(2486).getChannel());
    }

    @Test
    public void testGetWiFiChannelByFrequencyNotFound() {
        assertEquals(WiFiChannel.UNKNOWN, fixture.getWiFiChannelByFrequency(2399));
        assertEquals(WiFiChannel.UNKNOWN, fixture.getWiFiChannelByFrequency(2409));
        assertEquals(WiFiChannel.UNKNOWN, fixture.getWiFiChannelByFrequency(2481));
        assertEquals(WiFiChannel.UNKNOWN, fixture.getWiFiChannelByFrequency(2481));
        assertEquals(WiFiChannel.UNKNOWN, fixture.getWiFiChannelByFrequency(2487));
        assertEquals(WiFiChannel.UNKNOWN, fixture.getWiFiChannelByFrequency(2500));
    }

    @Test
    public void testGetWiFiChannelByChannel() {
        assertEquals(2412, fixture.getWiFiChannelByChannel(1).getFrequency());
        assertEquals(2437, fixture.getWiFiChannelByChannel(6).getFrequency());
        assertEquals(2442, fixture.getWiFiChannelByChannel(7).getFrequency());
        assertEquals(2472, fixture.getWiFiChannelByChannel(13).getFrequency());
        assertEquals(2484, fixture.getWiFiChannelByChannel(14).getFrequency());
    }

    @Test
    public void testGetWiFiChannelByChannelNotFound() {
        assertEquals(WiFiChannel.UNKNOWN, fixture.getWiFiChannelByChannel(0));
        assertEquals(WiFiChannel.UNKNOWN, fixture.getWiFiChannelByChannel(15));
    }

    @Test
    public void testGetWiFiChannelFirst() {
        assertEquals(1, fixture.getWiFiChannelFirst().getChannel());
    }

    @Test
    public void testGetWiFiChannelLast() {
        assertEquals(14, fixture.getWiFiChannelLast().getChannel());
    }

    @Test
    public void testGetWiFiChannelPairs() {
        List<Pair<WiFiChannel, WiFiChannel>> pair = fixture.getWiFiChannelPairs();
        assertEquals(1, pair.size());
        validatePair(1, 14, pair.get(0));
    }

    @Test
    public void testGetWiFiChannelPair() {
        validatePair(1, 14, fixture.getWiFiChannelPairFirst(Locale.US.getCountry()));
        validatePair(1, 14, fixture.getWiFiChannelPairFirst(null));
    }

    private void validatePair(int expectedFirst, int expectedSecond, Pair<WiFiChannel, WiFiChannel> pair) {
        assertEquals(expectedFirst, pair.first.getChannel());
        assertEquals(expectedSecond, pair.second.getChannel());
    }

    @Test
    public void testGetAvailableChannels() {
        assertEquals(11, fixture.getAvailableChannels(Locale.US.getCountry()).size());
        assertEquals(13, fixture.getAvailableChannels(Locale.UK.getCountry()).size());
    }

    @Test
    public void testGetWiFiChannelByFrequency2GHZ() {
        // setup
        Pair<WiFiChannel, WiFiChannel> wiFiChannelPair = fixture.getWiFiChannelPairs().get(0);
        // execute
        WiFiChannel actual = fixture.getWiFiChannelByFrequency(2000, wiFiChannelPair);
        // validate
        assertEquals(WiFiChannel.UNKNOWN, actual);
    }

    @Test
    public void testGetWiFiChannelByFrequency2GHZInRange() {
        // setup
        Pair<WiFiChannel, WiFiChannel> wiFiChannelPair = fixture.getWiFiChannelPairs().get(0);
        // execute
        WiFiChannel actual = fixture.getWiFiChannelByFrequency(wiFiChannelPair.first.getFrequency(), wiFiChannelPair);
        // validate
        assertEquals(wiFiChannelPair.first, actual);
    }

}

File: app/src/test/kotlin/com/vrem/wifianalyzer/wifi/band/WiFiChannelsGHZ2Test.kt
/*
 * WiFiAnalyzer
 * Copyright (C) 2015 - 2024 VREM Software Development <VREMSoftwareDevelopment@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>
 */
package com.vrem.wifianalyzer.wifi.band

import com.vrem.util.EMPTY
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test
import java.util.Locale

class WiFiChannelsGHZ2Test {
    private val fixture: WiFiChannelsGHZ2 = WiFiChannelsGHZ2()

    @Test
    fun inRange() {
        assertThat(fixture.inRange(2400)).isTrue()
        assertThat(fixture.inRange(2499)).isTrue()
    }

    @Test
    fun notInRange() {
        assertThat(fixture.inRange(2399)).isFalse()
        assertThat(fixture.inRange(2500)).isFalse()
    }

    @Test
    fun wiFiChannelByFrequency() {
        assertThat(fixture.wiFiChannelByFrequency(2410).channel).isEqualTo(1)
        assertThat(fixture.wiFiChannelByFrequency(2412).channel).isEqualTo(1)
        assertThat(fixture.wiFiChannelByFrequency(2414).channel).isEqualTo(1)
        assertThat(fixture.wiFiChannelByFrequency(2437).channel).isEqualTo(6)
        assertThat(fixture.wiFiChannelByFrequency(2442).channel).isEqualTo(7)
        assertThat(fixture.wiFiChannelByFrequency(2470).channel).isEqualTo(13)
        assertThat(fixture.wiFiChannelByFrequency(2472).channel).isEqualTo(13)
        assertThat(fixture.wiFiChannelByFrequency(2474).channel).isEqualTo(13)
        assertThat(fixture.wiFiChannelByFrequency(2482).channel).isEqualTo(14)
        assertThat(fixture.wiFiChannelByFrequency(2484).channel).isEqualTo(14)
        assertThat(fixture.wiFiChannelByFrequency(2486).channel).isEqualTo(14)
    }

    @Test
    fun wiFiChannelByFrequencyNotFound() {
        assertThat(fixture.wiFiChannelByFrequency(2399)).isEqualTo(WiFiChannel.UNKNOWN)
        assertThat(fixture.wiFiChannelByFrequency(2409)).isEqualTo(WiFiChannel.UNKNOWN)
        assertThat(fixture.wiFiChannelByFrequency(2481)).isEqualTo(WiFiChannel.UNKNOWN)
        assertThat(fixture.wiFiChannelByFrequency(2481)).isEqualTo(WiFiChannel.UNKNOWN)
        assertThat(fixture.wiFiChannelByFrequency(2487)).isEqualTo(WiFiChannel.UNKNOWN)
        assertThat(fixture.wiFiChannelByFrequency(2500)).isEqualTo(WiFiChannel.UNKNOWN)
    }

    @Test
    fun wiFiChannelByChannel() {
        assertThat(fixture.wiFiChannelByChannel(1).frequency).isEqualTo(2412)
        assertThat(fixture.wiFiChannelByChannel(6).frequency).isEqualTo(2437)
        assertThat(fixture.wiFiChannelByChannel(7).frequency).isEqualTo(2442)
        assertThat(fixture.wiFiChannelByChannel(13).frequency).isEqualTo(2472)
        assertThat(fixture.wiFiChannelByChannel(14).frequency).isEqualTo(2484)
    }

    @Test
    fun wiFiChannelByChannelNotFound() {
        assertThat(fixture.wiFiChannelByChannel(0)).isEqualTo(WiFiChannel.UNKNOWN)
        assertThat(fixture.wiFiChannelByChannel(15)).isEqualTo(WiFiChannel.UNKNOWN)
    }

    @Test
    fun wiFiChannelFirst() {
        assertThat(fixture.wiFiChannelFirst().channel).isEqualTo(1)
    }

    @Test
    fun wiFiChannelLast() {
        assertThat(fixture.wiFiChannelLast().channel).isEqualTo(14)
    }

    @Test
    fun wiFiChannelPairs() {
        val pair: List<WiFiChannelPair> = fixture.wiFiChannelPairs()
        assertThat(pair).hasSize(1)
        validatePair(pair[0])
    }

    @Test
    fun wiFiChannelPair() {
        validatePair(fixture.wiFiChannelPairFirst(Locale.US.country))
        validatePair(fixture.wiFiChannelPairFirst(String.EMPTY))
    }

    private fun validatePair(pair: WiFiChannelPair) {
        assertThat(pair.first.channel).isEqualTo(1)
        assertThat(pair.second.channel).isEqualTo(14)
    }

    @Test
    fun availableChannels() {
        assertThat(fixture.availableChannels(Locale.US.country)).hasSize(11)
        assertThat(fixture.availableChannels(Locale.UK.country)).hasSize(13)
    }

    @Test
    fun wiFiChannelByFrequency2GHZ() {
        // setup
        val wiFiChannelPair: WiFiChannelPair = fixture.wiFiChannelPairs()[0]
        // execute
        val actual: WiFiChannel = fixture.wiFiChannelByFrequency(2000, wiFiChannelPair)
        // validate
        assertThat(actual).isEqualTo(WiFiChannel.UNKNOWN)
    }

    @Test
    fun wiFiChannelByFrequency2GHZInRange() {
        // setup
        val wiFiChannelPair: WiFiChannelPair = fixture.wiFiChannelPairs()[0]
        // execute
        val actual: WiFiChannel = fixture.wiFiChannelByFrequency(wiFiChannelPair.first.frequency, wiFiChannelPair)
        // validate
        assertThat(actual).isEqualTo(wiFiChannelPair.first)
    }
}